submissions/angels_unit_tests/angels_unit_tests.md

```diff
@@ -0,0 +1,620 @@
+
+We realised that with the overhauls we created many more configuration, with even more settings. At this point it becomes quite tough to maintain and test this completely by hand. I started thinking about creating some [unit tests](https://en.wikipedia.org/wiki/Unit_testing) to take this burden of our shoulders. After some discussions and ideas from the great modding community I finally decided on the 30th of June 2021 to commit myself at creating the unit test infrastructure, groupted in a separate mod: [Angel's Dev - Unit Tests](https://github.com/Arch666Angel/mods/pull/634/commits/d59328e81235e26beb88ba66659cb78315827a00). A small note, You will not find this mod on the modportal.
+
+I already had a way of building the Angel's mods from the github folder to the factorio mod folder and a similar script to download bob mods from the mod portal. In order to test different configurations, I had to write two additional scripts. [The first script](https://github.com/Arch666Angel/mods/pull/634/commits/8977e17bba55d06fb3b60ff4c593e2307f110395) reads and rewrites the `mod-settings.dat` file with the mod settings. The settings file stores all the user configured settings, which can lead to different mod configurations. For example, in Angel's Industries it can be used to switch between *special vanilla*, *bob-angels* (= regular overhaul, also possible without bob mods), *component overhaul* or *science overhaul*. It basically can change the complete behaviour of angels mods with a few checkboxes:
+
+![Settings within Angel's Industries](https://raw.githubusercontent.com/LovelySanta/Alt-F4/angels-unit-tests/submissions/angels_unit_tests/angels_industry_settings.png)
+
+Since the loading and saving part of mod settings is completely handled by the base game, it was a part that I didn't encounter in creating mods before. However, I was very surprised at finding [thorough documentation on the Factorio Wiki](https://wiki.factorio.com/Mod_settings_file_format) about this.
+
+[The second script](https://github.com/Arch666Angel/mods/pull/634/commits/abb64c2fe8007e060f112716ebe33864510b6451) configures the `mod-list.json` file, which contains the information about which mods have to be enabled. This is the part which are people most familiar with that play modded Factorio. They download mods from the portal, and after finishing the game they disable the mods and play with some other mods.
+
+![Factorio mod list with all Bob Angel's mods](https://raw.githubusercontent.com/LovelySanta/Alt-F4/angels-unit-tests/submissions/angels_unit_tests/mod_list.png)
```

Comment:
Could you expand your caption to add that this was all set with your script?

Reply:
Resolved in https://github.com/AlternativeFFFF/Alt-F4/pull/551/commits/09e0500a041063c47fe716375139b83f823eb7e9.